src/main/frontend/components/select.cljs

```diff
@@ -0,0 +1,105 @@
+  (:require [frontend.modules.shortcut.core :as shortcut]
+            [frontend.context.i18n :as i18n]
+            [frontend.search :as search]
+            [frontend.state :as state]
+            [frontend.ui :as ui]
+            [frontend.util :as util]
+            [frontend.db :as db]
+            [frontend.text :as text]
+            [rum.core :as rum]
+            [frontend.config :as config]
+            [reitit.frontend.easy :as rfe]))
+
+(rum/defc render-item
+  [{:keys [id value]} chosen?]
+  [:div.inline-grid.grid-cols-4.gap-x-4.w-full
+   {:class (when chosen? "chosen")}
+   [:span.col-span-3 value]
+   [:div.col-span-1.justify-end.tip.flex
+    (when id
+      [:code.opacity-20.bg-transparent id])]])
+
+(rum/defcs select <
+  (shortcut/disable-all-shortcuts)
+  (rum/local "" ::input)
+  {:will-unmount (fn [state]
+                   (state/set-state! [:ui/open-select] nil)
+                   state)}
+  [state {:keys [items limit on-chosen empty-placeholder prompt-key]
+          :or {limit 100
+               prompt-key :select/default-prompt}}]
+  (rum/with-context [[t] i18n/*tongue-context*]
+    (let [input (::input state)]
+      [:div.cp__select.cp__select-main
+       [:div.input-wrap
+        [:input.cp__select-input.w-full
+         {:type        "text"
+          :placeholder (t prompt-key)
+          :auto-focus  true
+          :value       @input
+          :on-change   (fn [e] (reset! input (util/evalue e)))}]]
+
+       [:div.item-results-wrap
+        (ui/auto-complete
+         (search/fuzzy-search items @input :limit limit :extract-fn :value)
+         {:item-render render-item
+          :class       "cp__select-results"
+          :on-chosen   (fn [x]
+                         (state/close-modal!)
+                         (on-chosen x))
+          :empty-placeholder (empty-placeholder t)})]])))
+
+(defn select-config
+  "Config that supports multiple types (uses) of this component. To add a new
+  type, add a key with the value being a map with the following keys:
+
+  * :items-fn - fn that returns items with a :value key that are used for the
+    fuzzy search and selection. Items can have an optional :id and are displayed
+    lightly for a given item.
+  * :on-chosen - fn that is given item when it is chosen.
+  * :empty-placeholder - fn that returns hiccup html to render if no matched graphs found.
+  * :prompt-key - dictionary keyword that prompts when components is first open.
+    Defaults to :select/default-prompt."
+  []
+  {:select-graph
+   {:items-fn (fn []
+                (->>
+                 (state/get-repos)
+                 (remove (fn [{:keys [url]}]
+                           (or (config/demo-graph? url)
+                               (= url (state/get-current-repo)))))
+                 (map (fn [{:keys [url]}]
+                        {:value (text/get-graph-name-from-path
+                                 ;; TODO: Use helper when a common one is refactored
```

Comment:
I would've like to have extracted [this helper](https://github.com/logseq/logseq/blob/dc2258fb76c997cd67a0b92ee6ba4d3f3f7e3384/src/main/frontend/components/repo.cljs#L194) and used it but it's not clear if this helper should belong in text, config or db and if we put them in one, it adds to the confusion by having them depend on each other. Also found it strange I couldn't reuse the result of the if block and get-repo-dir since they share most of the same implentation. I don't mind cleaning up these helpers as I understand them better

Reply:
Worth noting by using the same helpers as components.repo, the items that display in this component are consistent with how graphs are displayed elsewhere to include parent dir e.g. 'foo/bar' instead of 'bar'